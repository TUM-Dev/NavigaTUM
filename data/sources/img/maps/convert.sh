#!/bin/bash
set -euo pipefail



# Step 1: convert geotiff to mbtiles

for tif in ./overlays/*_modified.tif; do
    # Strip extension and replace with .mbtiles
    mbtiles="${tif%_modified.tif}.mbtiles"

    echo "Processing $tif -> $mbtiles"
    rio mbtiles \
        --format WEBP \
        --zoom-levels 16..20 \
        --resampling lanczos \
        --rgba \
        --progress-bar \
        --exclude-empty-tiles \
        "./$tif" "./$mbtiles"
done

# Move to different directory
rm -f overlay_levels/*
mv overlays/*.mbtiles overlay_levels/

# Step 2: Merge per level
levels=$(ls ./overlay_levels/*.mbtiles | sed -E 's/.*_(.+)\.mbtiles/\1/' | sort -u)
echo "Merging mbtiles by levels: $levels"

for lvl in $levels; do
    echo "Processing level $lvl"
    input="$(ls ./overlay_levels/*_$lvl.mbtiles)"
    sources=$(echo "$input" | tr '\n' ',' | sed 's#./overlay_levels/##g' | sed 's#\.mbtiles##g' | sed 's#,$##')
    output="./overlay_levels/${lvl}.mbtiles"
    martin-cp $(echo "$input" | tr '\n' ' ') --source "$sources" --auto-bounds calc --encoding identity --max-zoom 20 --min-zoom 16 --mbtiles-type normalized --output-file "${output}"
    rm $(echo "$input" | tr '\n' ' ')
done
