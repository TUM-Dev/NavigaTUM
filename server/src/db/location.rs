use chrono::{DateTime, Utc};
use sqlx::PgPool;

#[allow(dead_code)] // used for testing out the repo pattern
#[derive(Debug)]
pub struct Location {
    pub last_calendar_scrape_at: Option<DateTime<Utc>>,
    pub lat: f64,
    pub lon: f64,
    pub name: String,
    pub type_common_name: String,
    pub r#type: String,
    pub calendar_url: Option<String>,
    pub tumonline_room_nr: Option<i32>,
    pub coordinate_accuracy: Option<String>,
    pub coordinate_source: String,
    pub comment: Option<String>,
    pub usage_id: Option<i32>,
    pub operator_id: Option<i32>,
}
impl Location {
    #[tracing::instrument(skip(pool))]
    pub async fn fetch_optional(
        pool: &PgPool,
        id: &str,
        should_use_english: bool,
    ) -> sqlx::Result<Option<Self>> {
        if should_use_english {
            sqlx::query_as!(
                Self,
                r#"SELECT last_calendar_scrape_at,lat,lon,name,type_common_name,type,calendar_url,tumonline_room_nr,coordinate_accuracy,coordinate_source,comment,usage_id,operator_id
                FROM en
                WHERE key=$1"#,
                id)
                .fetch_optional(pool).await
        } else {
            sqlx::query_as!(
                Self,
                r#"SELECT last_calendar_scrape_at,lat,lon,name,type_common_name,type,calendar_url,tumonline_room_nr,coordinate_accuracy,coordinate_source,comment,usage_id,operator_id
                FROM de
                WHERE key=$1"#,
                id)
                .fetch_optional(pool).await
        }
    }
}

#[derive(Debug, Clone)]
pub struct LocationKeyAlias {
    pub key: String,
    pub visible_id: String,
    pub r#type: String,
}
impl LocationKeyAlias {
    pub async fn fetch_optional(pool: &PgPool, id: &str) -> sqlx::Result<Option<Self>> {
        sqlx::query_as!(
            Self,
            r#"
        SELECT key, visible_id, type
        FROM aliases
        WHERE alias = $1 AND key <> alias
        LIMIT 1"#,
            id
        )
        .fetch_optional(pool)
        .await
    }
}
