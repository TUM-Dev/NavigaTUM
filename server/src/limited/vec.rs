use std::fmt;
use std::vec::IntoIter;

use serde::{Deserialize, Serialize};

use crate::limited::OrMore;

#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct LimitedVec<T>(pub Vec<T>);

impl<T> AsRef<[T]> for LimitedVec<T> {
    fn as_ref(&self) -> &[T] {
        &self.0
    }
}

impl<T> IntoIterator for LimitedVec<T> {
    type Item = T;
    type IntoIter = IntoIter<T>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl<T> LimitedVec<T> {
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
    pub fn len(&self) -> usize {
        self.0.len()
    }
    pub fn pop(&mut self) -> Option<T> {
        self.0.pop()
    }
}

impl<T> From<Vec<T>> for LimitedVec<T> {
    fn from(value: Vec<T>) -> Self {
        LimitedVec(value)
    }
}

const LIMIT: usize = 3;
impl<T: fmt::Debug> fmt::Debug for LimitedVec<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.0.len() <= LIMIT {
            f.debug_list().entries(self.0.iter().take(LIMIT)).finish()
        } else {
            f.debug_list()
                .entries(
                    self.0
                        .iter()
                        .take(LIMIT)
                        .map(OrMore::Value)
                        .chain([OrMore::More]),
                )
                .finish()
        }
    }
}
impl<T> FromIterator<T> for LimitedVec<T> {
    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
        let mut c = Vec::new();

        for i in iter {
            c.push(i);
        }

        LimitedVec(c)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_limited_output() {
        let w: LimitedVec<u32> = LimitedVec(vec![]);
        assert_eq!(format!("{w:?}"), "[]");
        let w = LimitedVec(vec![1]);
        assert_eq!(format!("{w:?}"), "[1]");
        let w = LimitedVec(vec![1, 2]);
        assert_eq!(format!("{w:?}"), "[1, 2]");
        let w = LimitedVec(vec![1, 2, 3]);
        assert_eq!(format!("{w:?}"), "[1, 2, 3]");
        let w = LimitedVec(vec![1, 2, 3, 4]);
        assert_eq!(format!("{w:?}"), "[1, 2, 3, ...]");
        let w = LimitedVec(vec![1, 2, 3, 4, 5]);
        assert_eq!(format!("{w:?}"), "[1, 2, 3, ...]");
    }
}
