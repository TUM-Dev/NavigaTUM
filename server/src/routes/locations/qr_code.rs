use crate::db::location::LocationKeyAlias;
use crate::limited::vec::LimitedVec;
use actix_web::http::header::{CacheControl, CacheDirective};
use actix_web::{HttpResponse, get, web};
use base64::Engine;
use base64::prelude::BASE64_STANDARD;
use fast_qr::convert::{Builder, Shape, image::ImageBuilder};
use fast_qr::qr::QRBuilder;
use serde::Deserialize;
use tracing::error;

#[derive(Deserialize, utoipa::IntoParams)]
struct QrCodePathParams {
    id: String,
}

#[tracing::instrument]
fn generate_qr_code(url: &str) -> anyhow::Result<LimitedVec<u8>> {
    // Build the QR code
    let qrcode = QRBuilder::new(url)
        .build()
        .map_err(|e| anyhow::anyhow!("Failed to build QR code: {e}"))?;

    // Generate the image with rounded corners
    let pin = include_bytes!("static/pin.png");
    let pin_base64 = BASE64_STANDARD.encode(pin);
    let png_code = ImageBuilder::default()
        .margin(1)
        .shape(Shape::RoundedSquare)
        .module_color("#3070B3")
        .background_color("#ffffff")
        .fit_width(600)
        .fit_height(600)
        .image(format!("data:image/png;base64,{pin_base64}"))
        .to_bytes(&qrcode)
        .map_err(|e| anyhow::anyhow!("cannot build QR code: {e}"))?;

    Ok(LimitedVec(png_code))
}

/// Get a QR code for a location
///
/// This returns a QR code image (PNG) that links to the location's detail page.
/// The QR code uses TUM blue (#0065bd) as foreground color with white background and rounded corners.
#[utoipa::path(
    tags=["locations"],
    params(QrCodePathParams),
    responses(
        (status = 200, description = "**QR code image**", content_type="image/png"),
        (status = 400, description = "**Bad request.** Make sure that requested item ID is not empty and not longer than 255 characters", body = String, content_type = "text/plain", example = "Invalid ID"),
        (status = 404, description = "**Not found.** Make sure that requested item exists", body = String, content_type = "text/plain", example = "Not found"),
        (status = 500, description = "**Internal server error**", body = String, content_type = "text/plain"),
    )
)]
#[get("/api/locations/{id}/qr-code")]
pub async fn qr_code_handler(
    params: web::Path<QrCodePathParams>,
    data: web::Data<crate::AppData>,
) -> HttpResponse {
    let id = params
        .id
        .replace(|c: char| c.is_whitespace() || c.is_control(), "");
    if params.id.is_empty() || params.id.len() > 255 {
        return HttpResponse::BadRequest()
            .content_type("text/plain")
            .body("Invalid ID");
    }

    let id = match LocationKeyAlias::fetch_optional(&data.pool, &id).await {
        Ok(Some(id)) => format!("https://nav.tum.de/{type}/{id}", type = id.r#type
          , id = id.key),
        Ok(None) => {
            return HttpResponse::NotFound()
                .content_type("text/plain")
                .body("Not found");
        }
        Err(e) => {
            error!(error = %e, "Failed to fetch location key alias. Assuming it is legitimate, since the generated links are a 404 in the worst case");
            format!("https://nav.tum.de/view/{id}")
        }
    };

    // Location exists, generate QR code
    match generate_qr_code(&id) {
        Ok(qr_image) => HttpResponse::Ok()
            .content_type("image/png")
            .insert_header(CacheControl(vec![
                CacheDirective::MaxAge(7 * 24 * 60 * 60), // valid for 7d
                CacheDirective::Public,
            ]))
            .body(qr_image.0),
        Err(e) => {
            error!(error = %e, "Failed to generate QR code");
            HttpResponse::InternalServerError()
                .content_type("text/plain")
                .body("Failed to generate QR code")
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_qr_code() {
        let result = generate_qr_code("https://nav.tum.de/view/5510.00.001");

        assert!(result.is_ok(), "QR code generation should succeed");
        let qr_image = result.unwrap();

        // Verify it's a valid PNG
        assert!(!qr_image.0.is_empty(), "QR code should not be empty");

        // PNG files start with these magic bytes
        assert_eq!(
            &qr_image.0[0..8],
            &[137, 80, 78, 71, 13, 10, 26, 10],
            "Should be a valid PNG file"
        );

        // Use insta binary snapshot to verify the exact output
        insta::assert_binary_snapshot!("qr_code_5510_00_001.png", qr_image.0);
    }

    #[test]
    /// Tests that the same ID always generates the same QR code
    fn test_qr_code_consistency() {
        let url = "https://nav.tum.de/view/mi";
        let result1 = generate_qr_code(url).unwrap();
        let result2 = generate_qr_code(url).unwrap();

        assert_eq!(
            result1.0, result2.0,
            "QR code generation should be deterministic"
        );
    }

    #[test]
    /// Test that different IDs generate different QR codes
    fn test_qr_code_different_ids() {
        let result1 = generate_qr_code("https://nav.tum.de/view/5510.01.001").unwrap();
        let result2 = generate_qr_code("https://nav.tum.de/view/5602.EG.001").unwrap();

        assert_ne!(
            result1.0, result2.0,
            "Different IDs should generate different QR codes"
        );
    }
}
