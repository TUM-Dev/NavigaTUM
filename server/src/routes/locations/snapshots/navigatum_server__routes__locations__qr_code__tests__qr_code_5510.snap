---
source: src/routes/locations/qr_code.rs
expression: qr_image.0
extension: 01.001.png
snapshot_kind: binary
---
